src/ucs/sys/string.c

```diff
@@ -325,49 +326,56 @@ const char* ucs_flags_str(char *buf, size_t max,
     return buf;
 }
 
-ssize_t ucs_path_calc_distance(const char *path1, const char *path2)
+size_t ucs_string_count_char(const char *str, char c)
 {
-    size_t i = 0;
-    unsigned distance;
-    char resolved_path1[PATH_MAX], resolved_path2[PATH_MAX];
-    char *larger_str;
-    size_t larger_len, smaller_len;
-    size_t rp_len1, rp_len2;
+    size_t len   = strlen(str);
+    size_t count = 0;
+    size_t i;
 
-    if ((NULL == realpath(path1, resolved_path1)) ||
-        (NULL == realpath(path2, resolved_path2))) {
-        return UCS_ERR_INVALID_PARAM;
+    for (i = 0; i < len; i++) {
+        if (str[i] == c) {
+            count++;
+        }
     }
 
-    rp_len1 = strlen(resolved_path1);
-    rp_len2 = strlen(resolved_path2);
+    return count;
+}
 
-    if (rp_len1 > rp_len2) {
-        larger_len  = rp_len1;
-        smaller_len = rp_len2;
-        larger_str  = resolved_path1;
-    } else {
-        larger_len  = rp_len2;
-        smaller_len = rp_len1;
-        larger_str = resolved_path2;
-    }
+size_t ucs_string_common_prefix_len(const char *str1, const char *str2)
+{
+    size_t i = 0;
+    size_t smaller_len;
+    size_t len1, len2;
+
+    len1        = strlen(str1);
+    len2        = strlen(str2);
+    smaller_len = (len1 > len2) ? len2 : len1;
 
     /* find offset at which the paths diverge excluding the terminating
        character */
-    while ((i < (smaller_len - 1)) &&
-           (resolved_path1[i] == resolved_path2[i])) {
+    while ((i < (smaller_len - 1)) && (str1[i] == str2[i])) {
         i++;
     }
 
-    distance = (resolved_path1[i] == resolved_path2[i]) ? 0 : 1;
+    return i;
+}
 
-    while (i < larger_len) {
-        if (larger_str[i] == '/') {
-            distance++;
-        }
-        i++;
+ssize_t ucs_path_calc_distance(const char *path1, const char *path2)
+{
+    size_t common_offset;
+    unsigned distance;
+    char rpath1[PATH_MAX], rpath2[PATH_MAX];
+
+    if ((NULL == realpath(path1, rpath1)) ||
+        (NULL == realpath(path2, rpath2))) {
+        return UCS_ERR_INVALID_PARAM;
     }
 
+    common_offset = ucs_string_common_prefix_len(rpath1, rpath2);
+    distance      = (rpath1[common_offset] == rpath2[common_offset]) ? 0 : 1;
```

Comment:
why needed?

Reply:
When difference appears in the middle of processing a segment, we would need to include that segment also in the number of segments that are different. If we only look for `/` characters from the point of mismatch, we'll skip one segment. For example `/abcd` and `/abxy` should return 1.

Reply:
right. so if (rpath1[common_offset] == rpath2[common_offset]) it actually means both have reached '\0'
i'd write it like this:
```
if (rpath1[common_offset] != rpath2[common_offset]) {
   ++distance;  /* count the differentiating path component */
}
```